bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+ *
+ *  Note: We support right-sided (or semi) inifinity, where start key can be set
+ * to any value: [x..NOT_SET); an empty end key implies semi-infinite range.
+ */
+RowRange RowRange::infinite_range() { return RowRange(); }
+
+/**
+ * Default one, right-open: [begin, end).
+ * Both start and end should be set.
+ */
+RowRange RowRange::right_open(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_closed(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_open(std::move(end));
+  return tmp;
+}
+
+/**
+ * open interval: (begin, end), both endpoints should be set.
+ */
+RowRange RowRange::open_interval(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_open(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_open(std::move(end));
+  return tmp;
+}
+
+// closed interval: [begin, end], both endpoints should be set.
+RowRange RowRange::closed_interval(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_closed(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_closed(std::move(end));
+  return tmp;
+}
+
+// left-open: (begin, end], both endpoints should be set.
+RowRange RowRange::left_open(std::string begin, std::string end) {
+  RowRange tmp;
+  tmp.range_.set_start_key_open(std::move(begin));
+  if (!end.empty())  // (Semi-)infinite range: end = "".
+    tmp.range_.set_end_key_closed(std::move(end));
+  return tmp;
+}
+
+/**
+ * Given any key 'prefix', the prefix range is given as the right-open interval:
+ *   ['prefix', successor('prefix')).
+ *
+ * When prefix is of the form "\xFF\xFF\xFF...\xFF", the successor is
+ * "" (empty string). In such a case, we set end key to 'prefix';
+ *
+ */
+RowRange RowRange::prefix_range(std::string prefix) {
+  RowRange tmp;
+  std::string next_prefix = Successor(prefix);
+
+  if (next_prefix.empty()) next_prefix = prefix;
+  tmp.range_.set_start_key_closed(std::move(prefix));
+  if (!next_prefix.empty())  // Infinite range for prefix="".
+    tmp.range_.set_end_key_open(std::move(next_prefix));
+  return tmp;
+}
+
+// An alias for right_open().
+RowRange RowRange::range(std::string begin, std::string end) {
+  return right_open(begin, end);
+}
+
+/*
+ * Increment the given key: gives the permutation string (in the key-space)
+ * that immediately follows the 'key'.
+ * Please note that, for a key of the form "\xFF\xFF\xFF...\xFF", the successor
+ * is "" (empty string).
+ */
+std::string RowRange::Successor(std::string const& key) {
+  if (key.empty()) {
+    return "";
+  }
+  std::size_t position = key.find_last_not_of('\xFF');
+
+  if (position == std::string::npos) {
+    return "";
+  }
+  const char last_char = key[position] + 1;
+  return (key.substr(0, position) + last_char);
+}
+
+/**
+ * A range  can be of folling types:
```

Comment:
This comment does not describe what the function does, the Doxygen comments should, consider:

```C++
/**
 * Return true if @p row is in this range.
 */  
```